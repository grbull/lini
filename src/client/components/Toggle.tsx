import React, { ChangeEvent, ReactElement } from 'react';

interface Props {
  id: string;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
  value: boolean;
}

// TODO: Need to wrap the input within the label for a11y,
// but I didn't want to fiddle with styles
export function Toggle({ id, onChange, value }: Props): ReactElement {
  return (
    <div className="relative inline-block w-10 mr-0 align-middle select-none transition duration-200 ease-in">
      <input
        checked={value}
        className="toggle-checkbox absolute block w-6 h-6 rounded-full bg-white border-4 appearance-none cursor-pointer"
        id={id}
        name={id}
        onChange={onChange}
        type="checkbox"
      />
      {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
      <label
        className="toggle-label block overflow-hidden h-6 rounded-full bg-gray-300 cursor-pointer"
        htmlFor={id}
      />
    </div>
  );
}
