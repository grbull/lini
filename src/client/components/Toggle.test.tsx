/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import { fireEvent, render, RenderResult } from '@testing-library/react';
import React from 'react';

import { Toggle } from './Toggle';

function setup(): RenderResult & {
  onChange: typeof jest.fn;
  input: HTMLInputElement;
} {
  const onChange = jest.fn();
  const utils = render(<Toggle id="test" onChange={onChange} value={false} />);

  const input = utils.getByLabelText('toggle') as HTMLInputElement;

  return { onChange, input, ...utils };
}

describe('Toggle Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });

  it('can be toggled', () => {
    const { onChange, input } = setup();

    fireEvent.click(input);

    expect(onChange).toBeCalled();
  });
});
