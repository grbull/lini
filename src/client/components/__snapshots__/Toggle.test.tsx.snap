// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative inline-block w-10 mr-0 align-middle select-none transition duration-200 ease-in"
  >
    <input
      aria-label="toggle"
      class="toggle-checkbox absolute block w-6 h-6 rounded-full bg-white border-4 appearance-none cursor-pointer"
      id="test"
      name="test"
      type="checkbox"
    />
    <label
      class="toggle-label block overflow-hidden h-6 rounded-full bg-gray-300 cursor-pointer"
      for="test"
    />
  </div>
</DocumentFragment>
`;
